src/librustc_trans/mir/constant.rs

```diff
@@ -1060,278 +135,240 @@ pub fn const_scalar_checked_binop<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
-                                vals: &[Const<'tcx>],
-                                discr: Option<Const<'tcx>>)
-                                -> Const<'tcx> {
-    assert_eq!(vals.len(), layout.fields.count());
 
-    match layout.abi {
-        layout::Abi::Scalar(_) |
-        layout::Abi::ScalarPair(..) |
-        layout::Abi::Vector { .. } if discr.is_none() => {
-            let mut non_zst_fields = vals.iter().enumerate().map(|(i, f)| {
-                (f, layout.fields.offset(i))
-            }).filter(|&(f, _)| !cx.layout_of(f.ty).is_zst());
-            match (non_zst_fields.next(), non_zst_fields.next()) {
-                (Some((x, offset)), None) if offset.bytes() == 0 => {
-                    return Const::new(x.llval, layout.ty);
-                }
-                (Some((a, a_offset)), Some((b, _))) if a_offset.bytes() == 0 => {
-                    return Const::new(C_struct(cx, &[a.llval, b.llval], false), layout.ty);
-                }
-                (Some((a, _)), Some((b, b_offset))) if b_offset.bytes() == 0 => {
-                    return Const::new(C_struct(cx, &[b.llval, a.llval], false), layout.ty);
-                }
-                _ => {}
+    pub fn mir_constant_to_miri_value(
+        &mut self,
+        bx: &Builder<'a, 'tcx>,
+        constant: &mir::Constant<'tcx>,
+    ) -> Result<MiriValue, ConstEvalErr<'tcx>> {
+        match constant.literal {
+            mir::Literal::Promoted { index } => {
+                let param_env = ty::ParamEnv::empty(traits::Reveal::All);
+                let cid = mir::interpret::GlobalId {
+                    instance: self.instance,
+                    promoted: Some(index),
+                };
+                bx.tcx().const_eval(param_env.and(cid))
             }
-        }
-        _ => {}
-    }
-
-    // offset of current value
-    let mut packed = false;
-    let mut offset = Size::from_bytes(0);
-    let mut cfields = Vec::new();
-    cfields.reserve(discr.is_some() as usize + 1 + layout.fields.count() * 2);
-
-    if let Some(discr) = discr {
-        let (field_size, field_align) = cx.size_and_align_of(discr.ty);
-        packed |= layout.align.abi() < field_align.abi();
-        cfields.push(discr.llval);
-        offset = field_size;
+            mir::Literal::Value { value } => {
+                Ok(self.monomorphize(&value))
+            }
+        }.and_then(|c| self.const_to_miri_value(bx, c))
     }
 
-    let parts = layout.fields.index_by_increasing_offset().map(|i| {
-        (vals[i], layout.fields.offset(i))
-    });
-    for (val, target_offset) in parts {
-        let (field_size, field_align) = cx.size_and_align_of(val.ty);
-        packed |= layout.align.abi() < field_align.abi();
-        cfields.push(padding(cx, target_offset - offset));
-        cfields.push(val.llval);
-        offset = target_offset + field_size;
+    // Old version of trans_constant now used just for SIMD shuffle
+    pub fn remove_me_shuffle_indices(&mut self,
```

Comment:
Now that it works you can rename this to something reasonable (and adjust the comment).

Reply:
done